blog: reducing tail latencies with auto yielding (#422)
tokio-rs/website: content/blog/2020-04-preemption.md

+++
date = "2020-03-31"
title = "Reducing tail latencies with automatic cooperative task yielding"
description = "April 1, 2020"
menu = "blog"
weight = 982
+++

Tokio is a runtime for asynchronous Rust applications. It allows writing code
using `async` & `await` syntax. For example:

```rust
let mut listener = TcpListener::bind(&addr).await?;

loop {
    let (mut socket, _) = listener.accept().await?;

    tokio::spawn(async move {
        // handle socket
    });
}
```

The Rust compiler transforms this code into a state machine. The Tokio runtime
executes these state machines, multiplexing many tasks on a handful of threads.
Tokio's scheduler requires that the generated task state machine yields control

> back to the scheduler in order to multiplex tasks. Each `.await` call is an

Reviewer: Note overlap with [this][1].

[1]: https://github.com/tokio-rs/website/pull/422#discussion_r401452196

opportunity to yield back to the scheduler. In the above example,
`listener.accept().await` will return a socket if one is pending. If there are
no pending sockets, control is yielded back to the scheduler.

This system works well in most cases. However, when a system comes under load,
it is possible for an asynchronous resource to never be "not ready". For
example, consider an echo server:

```rust
tokio::spawn(async move {
    let mut buf = [0; 1024];

    loop {
        let n = socket.read(&mut buf).await?;

        if n == 0 {
            break;
        }

        // Write the data back
        socket.write(buf[..n]).await?;
    }
});
```

If data is received faster than it can be processed, it is possible that, by the
time the processing of a data chunk completes, more data has already been
received. In this case, `.await` will never yield control back to the scheduler,
other tasks will not be scheduled, resulting in starvation and large latency
variance.

Currently, the answer to this problem is that the user of Tokio is responsible

> for adding yield points every so often. In practice, very few actually do this

Reviewer 2: Should we link to `yield_now`, and maybe also https://github.com/rust-lang/futures-rs/issues/2047 ?

and end up being vulnerable to this sort of problem.

> A common solution to this problem is preemption. OS threads will interrupt

Reviewer 2: "With normal OS threads, the kernel will interrupt..."

thread execution every so often in order to ensure fair scheduling of all
threads. Runtimes that have full control over execution (golang, erlang, ...)
will also use preemption to ensure fair scheduling of tasks. This is
accomplished by injecting yield points when compiling code that check if the
task has been executing for long enough and yielding back to the scheduler.

> Unfortunately, Tokio is not able to use this technique as `async` Rust does not

Reviewer 2: ```suggestion
Unfortunately, Tokio is not able to use this technique as Rust's `async` generators
do not provide any mechanism for executors (like Tokio) to inject such yield points.
```

> Unfortunately, Tokio is not able to use this technique as `async` Rust does not
> inject any sort of yield point.

Reviewer 2: ```suggestion
Unfortunately, Tokio is not able to use this technique as Rust's `async` generators
do not provide any mechanism for executors (like Tokio) to inject such yield points.
```

## Per-task operation budget

Even though Tokio is not able to **preempt**, there is still an opportunity to
nudge a task to yield back to the scheduler. As of [0.2.14], each Tokio task has
an operation budget. This budget is reset when the scheduler switches to the

> task. Each Tokio resource (socket, timer, channel, ...) is now aware of this

Reviewer 2: ```suggestion
task. Each Tokio resource (socket, timer, channel, ...) is aware of this
```

> task. Each Tokio resource (socket, timer, channel, ...) is now aware of this

Reviewer: I think you should have either "e.g." or "etc." here but not both.

> budget. As long as the task as budget remaining, the resource operates as it did

Reviewer 2: ```suggestion
budget. As long as the task has budget remaining, the resource operates as it did
```

previously. Each asynchronous operation (actions that users must `.await` on)

> decrement the task's budget. Once the task is out of budget, all resources will

Reviewer 2: This isn't really true though. It's only true if they `await` a tokio resource (the sentence says "Each asynchronous operation"). And I guess we also don't want to get into the details of how it's _really_ every `poll` call, not every `.await`.

Author: How would you update it... I guess I can say "all tokio resources".

> become "not ready" until the task yields back to the scheduler, at which point,
> the budget is reset.

Reviewer 2: ```suggestion
perpetually return "not ready" until the task yields back to the scheduler. At that point,
the budget is reset, and future `.await`s on Tokio resources will again function normally.
```

Reviewer 2: Do we also want to mention here that it will also call `.wake()` (said in different words perhaps) so that people don't go "won't this mean I might never be awoken again?!!)

Going back to the echo server example from above. When the task is scheduled, it

> is assigned a budget of 128 operations. When `socket.read(..)` and

Reviewer 2: This 128 sticks out a bit. May warrant a sentence of explanation.

Author: Say more? I don't think the exact number is important, but it seemed weird to write w/o specifying the number. I could sub it for `n`.

Reviewer 2: Ah, sorry, what I meant was that maybe we should have a sentence about why we chose 128, and how it's an arbitrary choice that will be revisited with more experience?

Reviewer 2: Or maybe even eventually dynamically adjusted using some heuristic?

the task yields back to the scheduler. If either `read` or `write` cannot
proceed due to the underlying socket not being ready (no pending data or a full
send buffer), then the task also yield back to the scheduler.

The idea originated from a conversation I was having with [Ryan Dahl][ry]. He is
using Tokio as the underlying runtime for [Deno][deno]. When doing some HTTP
experimentation (a while back) with [Hyper], he was seeing some high tail
latencies in some benchmarks. The problem was due to a loop not yielding back to
the scheduler under load. Hyper ended up fixing the problem by hand in this one
case, but Ryan mentioned that, when he worked on [node.js][node], they handled
the problem by adding **per resource** limits. So, if a TCP socket was always
ready, it would force a yield every so often. I mentioned this conversation to
[Jon Gjenset][jonhoo], and he ended up coming up with the idea of placing the
limit on the task itself instead of on each resource.

The end result is that Tokio should be able to provide more consistent runtime
behavior under load. While the exact heuristics will most likely be tweaked over

> time, initial measurements show that, in some cases, tail latencies are reduced
> almost 3x.

Reviewer: ```suggestion
time, initial measurements show that, in some cases, tail latencies are reduced
by almost 3x.
```

[![benchmark](https://user-images.githubusercontent.com/176295/73222456-4a103300-4131-11ea-9131-4e437ecb9a04.png)](https://user-images.githubusercontent.com/176295/73222456-4a103300-4131-11ea-9131-4e437ecb9a04.png)

"master" is before the automatic yielding and "preempt" is after. Click for a
bigger version, also see the original [PR comment][pr] for more details.

> 

Reviewer 2: There data I have on overhead is mainly the graph that's already included, and the lack of regression in noria.

> 

Author: I wouldn't worry too much about the lack of regression in this blog post.

## A note on blocking

While automatic cooperative task yielding improves many cases, it cannot preempt
tasks. Users of Tokio must still take care to avoid CPU intensive work and using
blocking APIs. The [`spawn_blocking`][spawn_blocking] function can be used to

> "asyncify" these sorts of tasks by running them on a thread pool.

Reviewer 2: Do we also want to mention `block_in_place` here?

Author: I left it out mostly because I would rather users default to `spawn_blocking`. I can add it though.

Reviewer 2: Either way is fine by me.

> While automatic cooperative task yielding improves many cases, it cannot preempt
> tasks. Users of Tokio must still take care to avoid CPU intensive work and using
> blocking APIs. The [`spawn_blocking`][spawn_blocking] function can be used to
> "asyncify" these sorts of tasks by running them on a thread pool.

Reviewer 3: nit: mention that it is a dedicated thread pool
```suggestion
Although automatic cooperative task yielding improves performance in many cases,
it cannot preempt tasks. Users of Tokio must still take care to avoid both CPU
intensive work and blocking APIs. The [`spawn_blocking`][spawn_blocking] function
can be used to "asyncify" these sorts of tasks by running them on dedicated a thread pool
where blocking is allowed.
```

Tokio does not, and will not attempt to detect blocking tasks and automatically
compensate by adding threads to the scheduler. This question has come up a
number of times in the past so allow me to elaborate.

For context, the idea is for the scheduler to include a monitoring thread. This
thread would poll scheduler threads every so often and check that workers are
making progress. If a worker is not making progress, it is assumed that the
worker is executing a blocking task and a new thread should be spawned to
compensate.

This idea is not new. The first occurence of this strategy that I am aware of is

> in the .NET thread pool and was introduced more than ten years ago.

Author: I can link this: https://mattwarren.org/2017/04/13/The-CLR-Thread-Pool-Thread-Injection-Algorithm/

> Unfortunately, the strategy has a number of problems and because of this, has
> not been featured in other thread pools / schedulers (golang, java, erlang,

Reviewer: ```suggestion
Unfortunately, the strategy has a number of problems and because of this, it has
not been featured in other thread pools / schedulers (golang, java, erlang,
```

...).

The first problem is that it is very hard to define "progress". A naive
definition of progress is whether or not a task has been scheduled for over some
unit of time. For example, if a worker has been stuck scheduling the same task
for more than 100ms, then that worker is flagged as blocked and a new thread is
spawned. In this definition, how does one detect scenarios where spawning a new
thread **reduces** throughput? This can happen when the scheduler is generally
under load and adding threads would make the situation much worse. To combat

> this, the .NET thread pool uses [hill climbing][hill].

Reviewer: I feel like a few more words can be added about the hill climbing heuristic they use?

Reviewer 4: I found https://mattwarren.org/2017/04/13/The-CLR-Thread-Pool-Thread-Injection-Algorithm/ which seems like a pretty good discussion of the specific hill-climbing approach used in CLR (at a glance). May be good as a second reference?

The second problem is that any automatic detection strategy will be vulnerable
to bursty or otherwise uneven workloads. This specific problem has been the bane

> of the .NET thread pool and is known as the "stuttering" problem. The hill

Reviewer 4: Are there references we can link to showing the impacts of the stuttering problem in .NET? After a quick google search, I found http://joeduffyblog.com/2006/07/08/clr-thread-pool-injection-stuttering-problems/ from Joe Duffy's blog (although it's also from 2006)

climbing strategy requires some period of time (hundreds of milliseconds) to
adapt to load changes. This time period is needed, in part, to be able to
determine that adding threads is improving the situation and not making it
worse.

The stuttering problem can be managed with the .NET thread pool, in part,
because the pool is designed to schedule **coarse** tasks, i.e. tasks that
execute in the order of hundreds of milliseconds to multiple seconds. However,

> asynchronous task schedulers are designed to schedule tasks that should run in

Reviewer 4: Isn't CLR also asynchronous or capable of asynchrony? Why was it not designed to schedule short tasks?

Should this just be 
```suggestion
in Rust, asynchronous task schedulers are designed to schedule tasks that should run in
```

> the order of micro seconds to tens of milliseconds at most. In this case, any

Reviewer 4: ```suggestion
the order of microseconds to tens of milliseconds at most. In this case, any
```

> stutttering problem from a heuristic based scheduler will result in far greater

Reviewer 4: nit:
```suggestion
stutttering problem from a heuristic-based scheduler will result in far greater
```

Reviewer 4: also, typo:
```suggestion
stuttering problem from a heuristic based scheduler will result in far greater
```

Reviewer: Overlap with [this](https://github.com/tokio-rs/website/pull/422#discussion_r401514725).

> the order of micro seconds to tens of milliseconds at most. In this case, any
> stutttering problem from a heuristic based scheduler will result in far greater
> latency variations.

Reviewer: ```suggestion
the order of microseconds to tens of milliseconds at most. In this case, any
stuttering problem from a heuristic based scheduler will result in far greater
latency variations.
```

> The most common follow up question I've received after this is "doesn't the Go

Reviewer 4: ```suggestion
The most common follow-up question I've received after this is "doesn't the Go
```

scheduler automatically detect blocked tasks?". The short answer is: no. Doing
so would result in the same stuttering problems as mentioned above. Also, Go has
no need to have generalized blocked task detection because Go is able to

> preempt. What the Go scheduler **does** do is annotate potentially blocking

Reviewer 4: Is there something we can link to for more information on how Go annotates potentially blocking calls?

Reviewer 2: Also, doesn't Go inject yield points as well? Good references here are https://github.com/golang/go/issues/10958 and https://github.com/golang/go/issues/24543.

Author: I mostly got this by reading the source...

Author: I'm not sure what I can ref.

system calls. This is roughly equivalent to the Tokio APIs

> [`spawn_blocking`][spawn_blocking] and [`block_in_place`][block_in_place].

Reviewer 4: The difference is that Go does this in the standard library, right?

Author: Tokio does as well... for example `tokio::fs`. The difference being that Tokio provides access to these fns as it doesn't preempt.

In short, as of now, the automatic cooperative task yielding strategy that has

> just been introduced is the best we have found for reducing tail latencies.

Reviewer 4: Should the conclusion maybe also have a sentence saying something about how, because tokio's resources now participate in automatic yielding, this will "just work" for most users without requiring them to do anything, but that the 'failure mode' (e.g. using `tokio::sync` on a runtime without automatic yielding) is just that everything continues working the way it was before?

<div style="text-align:right">&mdash;Carl Lerche</div>

> [0.2.14]: #

Reviewer: Remember to update this once it has been released.

Author: Thanks for reminding me, i had already forgotten... I'll probably forget again anyway 😆

stutttering problem from a heuristic-based scheduler will result in far greater
latency variations.

The most common follow-up question I've received after this is "doesn't the Go
scheduler automatically detect blocked tasks?". The short answer is: no. Doing
so would result in the same stuttering problems as mentioned above. Also, Go has
no need to have generalized blocked task detection because Go is able to
preempt. What the Go scheduler **does** do is annotate potentially blocking
system calls. This is roughly equivalent to the Tokio's
[`block_in_place`][block_in_place].

In short, as of now, the automatic cooperative task yielding strategy that has
just been introduced is the best we have found for reducing tail latencies.
Because this strategy only requires Tokio's types to opt-in, the end user does
not need to change anything to gain this benefit. Simply upgrading the Tokio
version will include this new functionality. Also, if Tokio's types are used
from **outside** of the Tokio runtime, they will behave as they did before.

There is more work that should happen on this topic. It is still how unclear how

> task budgets should work with "sub schedulers" (e.g.

Reviewer 4: nit/TIOLI:
```suggestion
task budgets should work with "sub-schedulers" (e.g.
```

[`FuturesUnordered`][futunord]). The task budget APIs should eventually be

> exposed so third party libs can integrate with them. It also would be nice to

Reviewer 4: ```suggestion
exposed publicly so that third party libs can integrate with them. It also would be nice to
```

figure out a way to generalize this concept so more than just Tokio users can

> take advantage of this.

Reviewer 4: ```suggestion
take advantage of it.
```

> take advantage of this.

Reviewer 4: Maybe the "future work" section could end with a "call to action"? "here's how you can help out if you're interested..." or even just "try it out by updating to today's tokio release"?

is accomplished by injecting yield points — code which checks if the task has
been executing for long enough and yields back to the scheduler if so — at
compile-time.  Unfortunately, Tokio is not able to use this technique as Rust's
`async` generators do not provide any mechanism for executors (like Tokio) to
inject such yield points.

## Per-task operation budget

Even though Tokio is not able to **preempt**, there is still an opportunity to
nudge a task to yield back to the scheduler. As of [0.2.14], each Tokio task has
an operation budget. This budget is reset when the scheduler switches to the
task. Each Tokio resource (socket, timer, channel, ...) is aware of this
budget. As long as the task has budget remaining, the resource operates as it did
previously. Each asynchronous operation (actions that users must `.await` on)
decrements the task's budget. Once the task is out of budget, all Tokio
resources will perpetually return "not ready" until the task yields back to the
scheduler. At that point, the budget is reset, and future `.await`s on Tokio
resources will again function normally.

Let's go back to the echo server example from above. When the task is scheduled, it

> is assigned a budget of 128 operations pr "tick". The number 128 was picked

Reviewer 5: `per`, not `pr`
